Added support in daily tasks for specifying ’every X days/weeks’

fallenpanda1 opened the pull request:
This is yet another attempt to implement the elusive "Repeat every X" feature: https://github.com/HabitRPG/habitrpg/issues/4173. I love habitrpg and wanted this feature really badly so I took some time to hack it out and also get a good feel for the code base.

Unfortunately, I noticed this other similar PR too late:
 https://github.com/HabitRPG/habitrpg/pull/4775.

As a result, I'm submitting this PR to see if you want me to continue working on this, or if you'd rather wait for the other PR. I recently just left my job as an iOS game dev, and I'm totally loving habitrpg, so I have plenty of time in the next few weeks to help out, either by finishing this PR or by helping with other features. If you want me to continue, I have designs planned out that I can share for monthly tasks as well as allowing a weekly/multi-day task to be completed on any day before the deadline (right now the way I've implemented it, it is greyed out until the exact specified date), definitely need your input. I'm also planning to clean up the code, fix the Date hackiness that you'll probably notice, and add unit tests.

Commit:
Added support in daily tasks for specifying ’every X days’ and ‘every X weeks on specific days of the week’.
- shouldDo() now takes the entire task as an input instead of just the ‘repeat’ (days of the week) dictionary.
- the ‘start’ date for a task can be specified (and can also be in the future, in which case it will be greyed out until then.) ‘start’ date also affects ‘every X days’ and ‘every X weeks’.
- no migration code yet.

fallenpanda1 commented:
Oh, this branch is functional, aside from the fact that I haven't added the migration code for existing tasks. E.g. you'll notice that an 'every 2 day' task will be greyed out every other day, depending on when you put the start date as. Here are some screen shots:
![screen shot 2015-05-08 at 12 29 19 pm](https://cloud.githubusercontent.com/assets/7478261/7543982/e7b5fb9e-f57d-11e4-95bd-4e51a13c336a.png)

![screen shot 2015-05-08 at 12 25 57 pm](https://cloud.githubusercontent.com/assets/7478261/7543951/aa111b7a-f57d-11e4-92f2-dec7be95942b.png)
![screen shot 2015-05-08 at 12 25 27 pm](https://cloud.githubusercontent.com/assets/7478261/7543950/aa0fddb4-f57d-11e4-82c6-7847d63cc513.png)

crookedneighbor commented:
This is a very good start, and I like the general direction you're taking with this.

Since your implementation changes the existing behavior of dailies, we're going to need to adjust the tests, and add new ones, to cover this new behavior. 

One of the tests for perfect days is now failing, so not only will we need to rewrite that test, but we'll have to be sure that any changes you made did not interfere with that feature (and if they did, do some rewriting so it works).

Not sure why the one of the group api tests is failing, so we'll have to investigate that.

crookedneighbor commented on common/script/index.coffee:
```diff
@@ -72,13 +72,51 @@ api.daysSince = (yesterday, options = {}) ->
+api.shouldDo = (day, dailyTask, options = {}) ->
+  return false unless dailyTask.type == 'daily' && dailyTask.repeat
+  # MIGRATION HACK, FIXME
+  if dailyTask.startDate == null
```
I like this, actually. A daily in this new format should always have a start date, right? 

Can we make it `if !dailyTask.startDate` instead to cover all falsy dates?

crookedneighbor replied:
We'll want to remove the `console.log`, of course.

fallenpanda1 replied:
Yup, a daily would always have a start date. I noticed that there's already a "creationDate" for each task, but it seems the creation date is meant more for recordkeeping of the actual creation date than something the user might change, hence my choice of adding this var. And oops! I'll remove the console.log!

fallenpanda1 replied:
Both comments done https://github.com/HabitRPG/habitrpg/commit/ba525eba61e273b1e730c0778b6a96158d41fd0b

crookedneighbor commented on common/script/index.coffee:
```diff
@@ -72,13 +72,51 @@ api.daysSince = (yesterday, options = {}) ->
+    console.log("null!!!!!")
+    dailyTask.startDate = moment().toDate();
+  if dailyTask.startDate instanceof String
+    console.log("startDate is a string: " + dailyTask.startDate + ". Converting to date");
```
I don't think this `console.log` is necessary here.

fallenpanda1 replied:
Removed https://github.com/HabitRPG/habitrpg/commit/ba525eba61e273b1e730c0778b6a96158d41fd0b.

crookedneighbor commented on common/script/index.coffee:
```diff
@@ -72,13 +72,51 @@ api.daysSince = (yesterday, options = {}) ->
+    return dayOfWeekCheck && everyXCheck && hasStartedCheck
+  else
+    # unexpected frequency string
+    return true
```
When it returns true, it displays as due, right? If it can't be calculated, would we want it to be displayed as due or gray? I'd vote gray, but am open to other opinions on the matter.

fallenpanda1 replied:
Hmm, are there error logging conventions? In this case, I think even more importantly than the default behavior, we want to wave some sort of yellow or red flag that this path was even reached. But yea, I agree that grey is probably the safer bet.

fallenpanda1 replied:
Now returns false https://github.com/fallenpanda1/habitrpg/commit/2028037d2ce5dcf17bdcbc77da499e7dffcc0bf6.

crookedneighbor commented on website/public/js/controllers/tasksCtrl.js:
```diff
@@ -129,6 +132,24 @@ habitrpg.controller("TasksCtrl", ['$scope', '$rootScope', '$location', 'User','N
+      task.startDate = task._dateString;
+    };
+
+    $scope.open = function($event) {
```
Couple of things
1. Let's name this something more specific. `$scope.openDatePicker` maybe?
2. Let's pass in the task as a second argument and set value for whether or not the datepicker should be open on the task itself. So:

```
$scope.openDatePicker = function($event, task) {
    $event.stopPropagation();
    task.datePickerOpen = !task.datePickerOpen
}
```

Then in the view, have the input appear on `task.datePickerOpen` instead.

The way you have it now, if multiple tasks are open in edit mode, the datepicker will appear for all of them if it's called on any one of them.

fallenpanda1 replied:
1. Sure. Good catch, open() is a pretty terrible name.
2. Ooh, got it! I'm a beginner at angular, so still making newbie mistakes.

fallenpanda1 replied:
1. and 2. both addressed https://github.com/HabitRPG/habitrpg/commit/2028037d2ce5dcf17bdcbc77da499e7dffcc0bf6

crookedneighbor commented on website/src/models/task.js:
```diff
@@ -50,10 +50,14 @@ var checklist = [{
-    type: {type:String, 'default': 'daily'},
+    type: {type: String, 'default': 'daily'},
+    //TODO: Cleaner to store interval as enum instead of str?
+    frequency: {type: String, 'default': 'daily'}, // 'daily', 'weekly'
```
Yes, I would store it as an enum.

crookedneighbor replied:
Default here is set to daily, but currently, a weekly is closer to what we have as a default. That means that 3rd party apps creating dailies, intending to create tasks that behave like weeklies will instead create tasks that behave like the new dailies format.

fallenpanda1 replied:
@ your first comment, what's the best way & place for me to declare this as an enum? Doesn't seem like js has traditional enums, and I'm not too familiar with conventions for js nor for habitrpg yet.

@ second comment, good call! I'm starting to understand that, in mongoose world, the properties of schemas are basically part of the public API, at least the way we use them here. I'll make the change.

fallenpanda1 replied:
Made the change https://github.com/fallenpanda1/habitrpg/commit/1af2fe33f733d000f5c6c5badabad17f67c000bf

crookedneighbor commented on website/views/shared/tasks/task.jade:
```diff
@@ -166,22 +166,56 @@ li(bindonce='list', id='task-{{::task.id}}', ng-repeat='task in obj[list.type+"s
+
+          // If frequency is daily
+          fieldset(ng-show='task.frequency=="daily"')
+            | Once every&nbsp;&nbsp;
```
We will want these to be translatable strings eventually, but don't worry about it now while we're hashing out exactly how this will work.

fallenpanda1 replied:
Hmm...just a thought, but I think we could pull this off relatively cleanly using mixins (thanks for pointing mixins out to me!):
The translation string will be "Once every <%= dropdown=%> days", which should be fairly standard for translators to deal with. For the jade, we write a mixin that takes in this string and splits with the marker as the delimiter:
mixin stringWithDropdownInBetween(string)
  // split string into 1st half before dropdown tag and 2nd half after the tag
  | {{firstHalf}}
  select
    ...
  | {{secondHalf}}

Quick disclaimer that I'm completely new to jade, so this might just be crazy talk.

fallenpanda1 replied:
Actually mixins aren't really necessary for my idea. Here's a simpler idea to take care of this case: https://github.com/fallenpanda1/habitrpg/commit/cc72c3bccabd0b55f2b4efdf1aef4ed6acc8251f. 

Don't really know the right delimiter to use here, and I'm also slightly concerned about performance from the use of regexes for string split.

crookedneighbor commented on website/views/shared/tasks/task.jade:
```diff
@@ -166,22 +166,56 @@ li(bindonce='list', id='task-{{::task.id}}', ng-repeat='task in obj[list.type+"s
+                button(ng-class='{active: task.repeat.f}', type='button', ng-click='task.challenge.id || (task.repeat.f= !task.repeat.f)') {{::moment.weekdaysMin(5)}}
+              li
+                button(ng-class='{active: task.repeat.s}', type='button', ng-click='task.challenge.id || (task.repeat.s = !task.repeat.s)') {{::moment.weekdaysMin(6)}}
+              | <br><br>
```
Would rather not have html in the jade files. Use css styles on the `.repeat-days` class if you need more of a gap.

fallenpanda1 replied:
Addressed https://github.com/HabitRPG/habitrpg/commit/2028037d2ce5dcf17bdcbc77da499e7dffcc0bf6.

crookedneighbor commented on website/views/shared/tasks/task.jade:
```diff
@@ -166,22 +166,56 @@ li(bindonce='list', id='task-{{::task.id}}', ng-repeat='task in obj[list.type+"s
+              option(ng-repeat='x in [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]') {{x}}
+            |   days<br><br>
+            | Start date:&nbsp;
+            p(class='input-group')
```
Is it important that the start date is nested inside the weekly/daily fieldsets? If not, we could move it out and not need to repeat the start date code.

fallenpanda1 replied:
Sure, I can move it out. Is there an easy way to repeat this snippet of code in both places? That seems like the best option, but I couldn't find any easy way of doing that. If not, then I'll just move it out.

crookedneighbor replied:
You could create a [jade mixin](http://jade-lang.com/reference/mixins/), but I think it would make more sense to move it into a separate fieldset.

crookedneighbor commented:
Another thing we need to consider is how to alert 3rd parties about changes to the API. Devs, any ideas? We could have a dedicated blog that posts upcoming changes to the api at a certain time before the change actually goes live. Maybe a week?

If the change is a breaking change, we should probably just make the change backwards compatible until v3 of the api.

Alys commented:
I'd like any changes to the API's the [documented behaviour](https://habitrpg.com/static/api) to be backwards compatible if at all possible. The discussion about how to notify third-party developers is in https://github.com/HabitRPG/habitrpg/issues/4432

But is this a problem for this PR? I haven't seen any changes to how the API would be used (except perhaps a new field that can be included in Dailies, but that would be backwards compatible, assuming it's not mandatory).

Commit:
Cleaned up task.jade file:
- moved ugly html newlines into the stylesheet
- renamed temporary var for datePicker’s ngmodel from _dateString to _tempDateForPicker
- renamed open -> openDatePicker, and have the function track datePicker’s open/closed state on a per task instead of global basis

Commit:
Removed stray debugging console logs.

Commit:
Set DailySchema’s ‘frequency’ default to weekly to maintain backwards compatibility.

fallenpanda1 commented:
@crookedneighbor K, took a pass at all your comments. A few questions/todos:
1. How do I run the unit tests? (I'm running using WebStorm, but getting any direction for running it is fine!)
2. An outstanding comment about localization, and one about how to make enums.

Also, is there any channel that devs commonly communicate using? I saw an IRC channel on the wiki, wasn't sure if that was it.

Commit:
Fixed issue where incorrect time type passed in for shouldDo’s ‘day’ argument (should be a Date, sometimes Moment is passed in). Fixed incorrect comparison between Moment and Date.

Commit:
In taskDefaults method, added default values for new fields: startDate, everyX, and frequency.

crookedneighbor commented on website/src/models/task.js:
```diff
@@ -50,10 +50,14 @@ var checklist = [{
-    type: {type:String, 'default': 'daily'},
+    type: {type: String, 'default': 'daily'},
+    //TODO: Cleaner to store interval as enum instead of str?
+    frequency: {type: String, 'default': 'weekly'}, // 'daily', 'weekly'
```
Here's an example of an enum: https://github.com/HabitRPG/habitrpg/blob/develop/website/src/models/coupon.js#L10

crookedneighbor commented:
@fallenpanda1 I used to keep the IRC channel open, but my current workplace blocks IRC channels, but I try to keep up with [the Aspiring Coders Guild](https://habitrpg.com/#/options/groups/guilds/68d4a01e-db97-4786-8ee3-05d612c5af6f).

crookedneighbor commented on common/script/index.coffee:
```diff
@@ -72,13 +72,50 @@ api.daysSince = (yesterday, options = {}) ->
+    dailyTask.startDate = moment().toDate();
+  if dailyTask.startDate instanceof String
+    #TODO: Unexpected code path reached. Log a warning.
+    dailyTask.startDate = moment(dailyTask.startDate).toDate();
```
If the server is handling these unexpected cases, why is there a need to log them? I'm open to changing my mind if you can give a good reason. It just doesn't seem like a big deal to me.

fallenpanda1 replied:
No, you're right. I forgot that this code is just on the client, so logging makes no sense.

crookedneighbor committed:
Adjust fallenPanda's PR

* Use enum for frequency in task model
* Adjust ui of task selection
* Correct check for hasStarted
* Remove semicolons
* Set default start day to today
* Correct perfect day and resting in inn tests to have dailies with start dates of a week ago
* Add tests for dailies/weeklies functionality

Alys commented:
I think there's a bug in the "Once every x days" feature. Here and now, it is 22 May. I have these settings which I think should make the Daily not due (grey), but it remains blue:

![screen shot 2015-05-22 at 8 19 55 pm](https://cloud.githubusercontent.com/assets/1495809/7768577/f56b53ea-00bf-11e5-82d7-a79363377a82.png)

Also, could the drop-down menus for choosing day and week numbers be changed to fields that allow any integer to be entered, so that we can have numbers larger than 10?

crookedneighbor commented:
Also, I've [fixed and un-pended the tests for dailies/cron](https://github.com/HabitRPG/habitrpg/blob/develop/test/common/algos.mocha.coffee#L778-L861) in `test/common/`. Your PR should default dailys to being weeklies that repeat every week (the current behavior of dailies), and therefore should pass the existing tests, but they don't. (I've kicked off the Travis build to illustrate that, it should fail)

My best guess is some timezone wonkiness that needs to be taken into account, though it's not apparent to me yet why that's happening. 

I also have made some adjustments (WIP), you can see them here: https://github.com/HabitRPG/habitrpg/commit/db7f45cb58b3c4d3dc2ef3dca92124850c1a804f

Alys commented:
The blue/grey shading problem from my previous comment has been fixed.

crookedneighbor commented:
@fallenpanda1 It'd be best to ask questions regarding your PR on your PR, rather than on a different PR :smile:

You said: 

> @crookedneighbor cool! I haven't seen anything checked in so I assume your changes are on your local machine. Should I be doing anything with the PR in the meanwhile? I've been figuring out the unit/behavior test system (got it running!), and I'd be happy to add tests covering the changes I made to shouldDo(). Or is there any other high priority stuff you would really like someone to work on?

If you have the habitrpg repo set up as a remote (I call mine  `upstream`), you can pull in the changes I made like this:

``` bash
git pull upstream blade-Finish_weeklies_pr
```

[If you can get the existing dailies tests to pass](https://github.com/HabitRPG/habitrpg/blob/develop/test/common/algos.mocha.coffee#L778-L861), I'm pretty sure we'll be able to merge this in. Right now, that's the main blocker.

fallenpanda1 commented:
Sweet, I'll try to get that resolved ASAP then!

Commit:
Removed unnecessary TODOs.

Commit:
Fixed algos test to pass in task instead of task.repeat to shouldDo (API changed a few commits ago, but missed updating this test).

Alys commented:
Once all the code changes are finished, I'd like to do some more manual testing before it's merged, even if all the automatic tests pass.

redphoenix and I have noticed a similar problem to the one I reported earlier. When the "weekly" option is chosen, deselecting today's day keeps the Daily blue, but deselecting yesterday's day turns it grey.

fallenpanda1 commented:
@Alys sure, I'll definitely take a look at that once I fix the tests. Thanks for the details on how to repro!

@crookedneighbor I just noticed that xit is broken...whenever you use xit, the ENTIRE describe block containing the xit fails to run. You can notice this in the latest develop branch: there are a few xits in both the "achievements" and "dailies" describe blocks, and because of those xits, none of the its in those blocks run at all. "dailies" tests are actually broken on develop as well once you remove the xits to get the rest of the its to run, though I'm working to fix those too right now.

Commit:
Fixed bug where we were using moment.startOf instead of api.startOfDay in shouldDo.

fallenpanda1 commented:
@crookedneighbor never mind I'm dumb :(. xit is working fine, there's just a weird quirk with it. What is happening is that whenever an it block is xit'ed, the entire describe block doesn't show up in the test results list for me, even though the tests are indeed being run. The only way they show up is if I make one of the it blocks fail (which is good...this quirk doesn't allow failures to go unnoticed). I'm not sure if this is just with my setup or if other people see this.

crookedneighbor committed:
Set daillys startDate in the past

crookedneighbor commented:
@fallenpanda1 While you're working on specific tests, I'd recommend changing the describe block from `describe 'foo', ->` to `describe.only 'foo', ->`. That way, when you run `mocha test/common` and the only tests that will run will be the tests in those blocks. 

After you made your most recent change, all that needed to be done to make those tests pass was add a start date in the past for the dailies. The tests are checking the status of them yesterday. If they have a startDate of today (the default, one of the changes I made), of course they won't be due yesterday! See this commit for details: https://github.com/crookedneighbor/habitrpg/commit/d3ecb885d4b4397dfab7a60d3dffcd3b8815427b

crookedneighbor commented:
Here's the commit on the HabitRPG branch, if you want to easily bring it in: https://github.com/HabitRPG/habitrpg/commit/d3ecb885d4b4397dfab7a60d3dffcd3b8815427b

crookedneighbor commented:
Good catch on using the `api.startOfDay` function. That takes into account timezone and custom day start. Going to do some additional manual testing, but I think this is very close.

Alys commented:
I'm getting odd behaviour on my Dailies this morning. Dailies that should be due today are grey. If I open one for editing, I see that it's set to the correct weekly settings (correct days of the week) and the start date is displayed as today, but it's greyed out. If I "edit" the start date to select today's date (i.e., the same date that was already showing) then the Daily stops being grey.

Commit:
Made sure to use api.startOfDay in dailies.coffee unit test when calculating day of the week.

fallenpanda1 commented:
FYI, I'm out till later tomorrow, but I'll be looking at fixing the remaining unit tests and Alys's comments as soon as I'm available again. Excited to hopefully get this in soon!

Alys commented:
I like how this works, and I'm looking forward to being able to use it, but I'm concerned about the complexity it introduces for new users and for those users who prefer simpler options. I'm also finding that "Weekly" doesn't feel like the intuitive option for the normal Daily selectors (even though when I think about it rationally it makes sense) and if I'm having that trouble, then other users probably will too.

I am wondering if it's worth leaving the previous UI elements in place as the default options, and then have a button or similar to replace them with the advanced elements.

Or perhaps have three buttons with longer, more descriptive text on them:
- "Repeat on specific weekdays" (the options that are currently live on the production site, and this would be the default)
- "Repeat every x days" (the "Daily" option from the screenshots above)
- "Repeat every x weeks" (the "Weekly" option from above)

Instead of or as well as that, we might want to put this on the beta site to allow all users to give us feedback about it before it becomes mandatory.

@lemoness @AnitaYCheng What do you think?

Alys commented:
There might also be problems with people using old versions of the mobile apps, possibly for weeks after this goes live. If they use the new options on the website to schedule their Dailies, but then use the apps most of the time when completing their Dailies, there might be combinations of the new options that could cause the Dailies to be due but to not show as due on the apps (leading to surprise damage from cron). We should look into whether that's possible, and if it is and we can't mitigate it, the Bailey announcement should perhaps warn about it - FYI @lemoness. (Note: It's after my bedtime, I haven't thought about this in detail, it could be nonsense, but I can't get to sleep until I get it out of  my head.)

fallenpanda1 commented:
> I'm also finding that "Weekly" doesn't feel like the intuitive option for the normal Daily selectors (even though when I think about it rationally it makes sense) and if I'm having that trouble, then other users probably will too.
> 
> I am wondering if it's worth leaving the previous UI elements in place as the default options, and then have a button or similar to replace them with the advanced elements.

@Alys I agree that "Weekly" with all the days selected isn't the intuitive option for Daily selectors. What if the default going forward is just "Daily" instead of "Weekly"? The only reason for having "Weekly" with all the days of the week selected is for migration from older versions. To add clarity we could also highlight only the current day of the week when the user selects the "Weekly" button. I'm also totally fine putting it in the advanced section, just let me know!

Meanwhile, the tests are passing on the server, but I'm suspicious that some tests are failing only during certain times of day due to our use of current time in the tests, so looking into that.

lemoness commented:
@alys - I also found the interface pretty confusing!  I think that your
idea of separating out the Repeat Every features is a good start. I wonder
if there's a way to make it even clearer. Would like to hear from
@anitaycheng and @veeeeeee.

On Sun, May 24, 2015 at 9:56 AM, Allen notifications@github.com wrote:

> I'm also finding that "Weekly" doesn't feel like the intuitive option for
> the normal Daily selectors (even though when I think about it rationally it
> makes sense) and if I'm having that trouble, then other users probably will
> too.
> 
> I am wondering if it's worth leaving the previous UI elements in place as
> the default options, and then have a button or similar to replace them with
> the advanced elements.
> 
> @Alys https://github.com/Alys I agree that "Weekly" with all the days
> selected isn't the intuitive option for Daily selectors. What if the
> default going forward is just "Daily" instead of "Weekly"? The only reason
> for having "Weekly" with all the days of the week selected is for migration
> from older versions. To add clarity we could also highlight only the
> current day of the week when the user selects the "Weekly" button. I'm also
> totally fine putting it in the advanced section, just let me know!
> 
> Meanwhile, the tests are passing on the server, but I'm suspicious that
> some tests are failing only during certain times of day due to our use of
> current time in the tests, so looking into that.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5172#issuecomment-105036685.

Alys commented on common/script/index.coffee:
```diff
@@ -72,13 +72,49 @@ api.daysSince = (yesterday, options = {}) ->
+    everyXCheck = (weeksSinceTaskStartWeek % dailyTask.everyX == 0)
+    return dayOfWeekCheck && everyXCheck && hasStartedCheck
+  else
+    # unexpected frequency string
```
I think this will cause a problem for any Daily that hasn't been created or edited with the new code (i.e., Dailies created on the current website, or the mobile apps, or third-party apps, or direct API calls). Dailies created with those methods could persist for months after this code goes live. This code needs to be fully backwards-compatible with the old style of Dailies.

Some of my Dailies do not have the `frequency` attribute, so the new defaults definitely aren't being applied to all Dailies.

Alys replied:
Some of my Dailies do not have the `frequency` attribute, so the new defaults definitely aren't being applied to all Dailies.

crookedneighbor replied:
The mongoose model for dailys should take care of that. It has defaults for all the new parameters, but I'll double check that.

crookedneighbor replied:
Interesting. For me, the defaults are only being coerced into existing dailies when the daily is saved. However, they're behaving as expected even without saving. 

Going to experiment a little with custom day start and see if that's what's mucking it up for others.

crookedneighbor commented:
So far, I'm finding that the everyX day feature is easy for me to wrap my head around and get results that I expect. The everyX weeks though is either completely wrong, or too confusing for me. Perhaps both.

I propose the following:
- we use the everyX day feature as it is here (swapping out the select box for a number input box)
- we use dailies as they are currently on production, with the addition of the day start feature so people can plan for future dailies, but do not include the every x week option
- Add an option for people to mark specific days of the month for things to be due. IE, every month on the 15th or every month on the last day of the month. Again, with the day start feature. 
- Move all this complicated stuff under the advanced options so that users who like simplicity can keep using the dailies as they are now, but users who want more control can fine tune it a little for their needs.

crookedneighbor commented:
Pinging @lemoness, @alys, @anitaycheng and @veeeeeee.

veeeeeee commented:
Seconding the feeling of complexity wrt the daily selector being classed under weekly. 

I agree with @Alys that having Weekly being the default selected is not beneficial from a user perspective. (The column is called Dailies....)

I also like @crookedneighbor's proposal, but I think I would need to play with it to provide further commentary.

crookedneighbor commented:
Some screenshots of my ui ideas:

![screen shot 2015-05-26 at 5 58 15 pm](https://cloud.githubusercontent.com/assets/2916945/7825355/a1d061f2-03d1-11e5-9aa7-585cd47fc8b0.png)

![screen shot 2015-05-26 at 5 58 02 pm](https://cloud.githubusercontent.com/assets/2916945/7825356/a7703bd2-03d1-11e5-9130-5c01f9605ca8.png)

![screen shot 2015-05-26 at 5 58 22 pm](https://cloud.githubusercontent.com/assets/2916945/7825357/ad01523e-03d1-11e5-9cd0-06b0898c0539.png)

After you've changed frequencies, the input for changing frequencies becomes available in the normal spot:
![screen shot 2015-05-26 at 6 05 24 pm](https://cloud.githubusercontent.com/assets/2916945/7825364/ce503ab8-03d1-11e5-8119-da925acce0e2.png)

veeeeeee commented:
What happens if you've decided on a frequency, and then decide later on to change it? E.g. changing from doing something on Tuesdays and Fridays, and then changing it to every 3 days.

janetmango commented:
I'm very excited for this feature and can't wait to try it, so will we get any kind of notification when it's ready to try in beta?

Will streaks still work and will the data still be accessible via Alys' Display Data Tool? I'm assuming so, since the only change is really how the frequency for the Dailies is calculated, right?  But I want to make sure that important functionality isn't lost.

crookedneighbor commented:
@veeeeeee Exactly what you'd expect. The days of the week would no longer be used to calculate whether it should be done, and instead, the everyX option would be used instead. If a user has gone a very long time since initially creating the daily, they may want to update the startDate to better make sure they have the calculation correct. 

Maybe it would be helpful to have a notice about the next day the task will be due? IE: "Daily is due today/tomorrow/in X days"

@janetmango Alys will probably need to update the DDT to work correctly with the new options. This will definitely be something we'll put on the beta site when it is ready to be tried out by real users to get feedback on the functionality. (@lemoness @veeeeeee correct me if I'm wrong about that)

veeeeeee commented:
@crookedneighbor That makes sense. But if a user has decided to change the frequency of the task and their starting point is your final image, where would someone go to make that adjustment?

crookedneighbor commented:
@veeeeeee Under the advanced options, I just didn't make the screenshot tall enough to show the advanced options link.

veeeeeee commented:
@crookedneighbor Aha, yes, that makes sense.

lemoness commented:
I really like this new UI layout! Makes it much cleaner.

Right now beta is testing a bunch of other stuff, so I'm not actually sure
if we'll test this there, but it would be nice. If we do, we'll post about
it in this thread for sure.

On Tue, May 26, 2015 at 5:07 PM, veeeeeee notifications@github.com wrote:

> @crookedneighbor https://github.com/crookedneighbor Aha, yes, that
> makes sense.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5172#issuecomment-105701373.

fallenpanda1 commented:
Just checking, should I go ahead and implement the new layout @crookedneighbor mocked up? Or is someone already doing that, or we're still deciding?

crookedneighbor commented:
@fallenpanda1 If you want to pull in the latest changes I've made:

``` bash
git pull upstream blade-Finish_weeklies_pr
```

crookedneighbor commented:
Feel free to make changes if you have ideas regarding the interface.

crookedneighbor commented:
@fallenpanda1 Any updates on your end? Everything is working pretty well for me, but I think there might still be some issues with people who have custom day starts.

crookedneighbor commented:
Here is what I am observing:

_Today_: Wed, June 3, 8:00 am

_Dailys:_
1. Weekly. Due every wed
2. Daily. Due ever 2 days starting on June 1

If I have a Custom Day Start (CDS) of 0 (so no CDS):
- everything works as intended. Both weekly and daily are due. 

If I have a CDS of 5am (so, before my current time):
- Weekly is due, but daily is not due

If I have a CDS of 11pm (so, after my current time):
- Weekly is not due, but daily is due

The behavior of the weekly is consistent with the current behavior of the site. It's my opinion that the behavior of the daily should be consistent with the behavior of the weekly regarding CDS. 

@Alys @veeeeeee @lemoness are you seeing the same behavior?

fallenpanda1 commented:
@crookedneighbor I've pulled in your latest changes on my local machine, but got a little sidetracked and haven't had time to look deeply into it yet. I'll do that in the next day or two and maybe start looking at some of the issues mentioned above. Would appreciate any help with the CDS issue or any other outstanding things!

crookedneighbor commented:
@fallenpanda1 I [pushed a new change today](https://github.com/HabitRPG/habitrpg/commit/705e0dce715c4d93e51ce13c35eb21e298ddc6f8) that I _think_ fixed the CDS problem.

crookedneighbor commented:
Also, just pushed this: https://github.com/HabitRPG/habitrpg/commit/076f7f0db971588f2b3117abdf453a42d1872532

Alys commented:
The "Restore Streak" section is no longer in the Advanced Options.

Alys commented:
The Restore Streak section is good now, thanks @crookedneighbor!

crookedneighbor commented:
@fallenpanda1 We're just doing final testing now. I'll check back here when it's ready to be merged in.

crookedneighbor commented:
Not sure on the timeline for release though, could be a while.

fallenpanda1 commented:
Cool, sounds awesome!

crookedneighbor commented:
@fallenpanda1 This will count towards a contributor tier! What's your User ID? You can find it in [Settings > API](https://habitrpg.com/#/options/settings/api). Be sure to only copy the User ID and _not_ your API Token.

fallenpanda1 commented:
It's 49b930d2-1d9c-4754-af42-0f3b1740c396, although I think you already found me? Thanks so much!

crookedneighbor commented:
Contributor tier awarded!
